### ISSUE TITLE
toolchain install: unchanged - (toolchain not installed)

### ISSUE THREAD

**Issue by Riateche** [rust-lang/rustup#1777]
**Problem**
Rustup failed to install a toolchain:
```
info: syncing channel updates for '1.34.0-x86_64-unknown-linux-gnu'
info: latest update on 2019-04-11, rust version 1.34.0 (91856ed52 2019-04-10)
info: downloading component 'rustc'
info: downloading component 'rust-std'
error: component download failed for rust-std-x86_64-unknown-linux-gnu
info: caused by: could not rename downloaded file from '/home/ri/.rustup/downloads/32e097e75f13975e24e45465f2911dd794194f2803e9f6ae96d6f6254596fa45.partial' to '/home/ri/.rustup/downloads/32e097e75f13975e24e45465f2911dd794194f2803e9f6ae96d6f6254596fa45'
info: caused by: No such file or directory (os error 2)
```
After that, it is not possible to retry the installation:
```
$ rustup toolchain install 1.34.0
info: syncing channel updates for '1.34.0-x86_64-unknown-linux-gnu'

  1.34.0-x86_64-unknown-linux-gnu unchanged - (toolchain not installed)

info: checking for self-updates
$ rustup toolchain uninstall 1.34.0
info: no toolchain installed for '1.34.0-x86_64-unknown-linux-gnu'
$ rustup update                                   
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: checking for self-updates
info: downloading self-update

  stable-x86_64-unknown-linux-gnu unchanged - rustc 1.34.0 (91856ed52 2019-04-10)
```
I managed to fix the error by deleting the `~/.rustup/update-hashes/1.34.0-x86_64-unknown-linux-gnu` file manually.

The expected behavior is that `rustup toolchain install 1.34.0` installs the toolchain if it's not installed. 

**Steps**
I don't know how to reproduce the exact installation failure I got, but it seems that you can just do the following:

1. `rustup toolchain install 1.34.0`
2. Remove `~/.rustup/toolchains/1.34.0-x86_64-unknown-linux-gnu`.
3. `rustup toolchain install 1.34.0`

**Notes**

Output of `rustup --version`:
```
rustup 1.17.0 (069c88ed6 2019-03-05)
```
Output of `rustup show`:
```
Default host: x86_64-unknown-linux-gnu

info: syncing channel updates for '1.34.0-x86_64-unknown-linux-gnu'
installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu
1.33.0-x86_64-unknown-linux-gnu

active toolchain
----------------

1.34.0-x86_64-unknown-linux-gnu (overridden by '<current_dir>/rust-toolchain')
(toolchain not installed)

```

**Comment by kinnison** [comment]
The behaviour when you remove the `~/.rustup/toolchains/FOO` is that the update-hash still exists in `~/.rustup/update-hashes/FOO` which causes the confusion.  I wonder if we should either remove the hash if the toolchain is not present, or would it be better to report that the toolchain was unexpectedly absent and require that the user run `rustup toolchain uninstall FOO` (which will currently fail to remove the hash) and fix uninstall to do the cleanup.

**Comment by kinnison** [comment]
I am prepared to help someone to get into `rustup` coding by working on this with them.

**Comment by Glamhoth** [comment]
> I am prepared to help someone to get into `rustup` coding by working on this with them.

I would be interested. How can I reach you?

**Comment by kinnison** [comment]
The best way is to join us on the `#wg-rustup` discord channel at https://discord.gg/cKdYFVz  (Note you'll have to jump through a hoop to be able to speak with us, check your direct messages when you join)

If you can't discord, then we can see if there's another way which works for both of us.

**Comment by kinnison** [comment]
(Hopefully) fixed in #1801
